File: src/application/use-cases/FirewallService.ts
import { Rule, type RuleMode, type RuleType } from '../../domain/Rule';
import { type RuleRepository } from '../ports/outbound/RuleRepository';
import { type MessagePublisher } from '../ports/outbound/MessagePublisher';
import { AppError, type NonEmptyList } from '../../domain/common/types';
import { config } from '../../main/env';

export type CreateRulesCommand = {
  type: RuleType;
  mode: RuleMode;
  values: NonEmptyList<string | number>;
};

export class FirewallService {
  private readonly queue = config.constants.QUEUES.FIREWALL_RULES;

  constructor(
    private readonly ruleRepository: RuleRepository,
    private readonly messagePublisher: MessagePublisher
  ) {}

  async addRules(command: CreateRulesCommand) {
    const { type, mode, values } = command;

    const createdRules = values.map(value => new Rule(undefined, type, mode, value, true));

    const envelope = this.messagePublisher.createEnvelope('CREATE_RULE', {
      type,
      mode,
      rules: createdRules.map(rule => rule.toJSON())
    });

    const responsePayload = await this.messagePublisher.publishRPC(this.queue, envelope);

    return {
      type,
      mode,
      values: responsePayload.rules,
      status: 'success'
    };
  }

  async deleteRules(ids: NonEmptyList<number>) {
    const existing = await this.ruleRepository.findByIds(ids);
    if (existing.length !== ids.length) {
      throw new AppError('RULE_NOT_FOUND', 'Specified rule ID not found.', 404);
    }

    const envelope = this.messagePublisher.createEnvelope('DELETE_RULE', { ids });
    await this.messagePublisher.publish(this.queue, envelope);

    return {
      removed: existing.map(rule => rule.toJSON()),
      status: 'success'
    };
  }

  async getRules(type?: string) {
    let normalizedType: RuleType | undefined;
    if (type) {
      const lower = type.toLowerCase();
      if (!['ip', 'domain', 'port'].includes(lower)) {
        throw new AppError('INVALID_TYPE_FILTER', 'Invalid rule type filter.');
      }
      normalizedType = lower as RuleType;
    }

    let rules = await this.ruleRepository.getAll();
    if (normalizedType) {
      rules = rules.filter(r => r.type === normalizedType);
    }

    const response: Record<string, Record<string, any>> = {
      ips: {},
      domains: {},
      ports: {}
    };

    rules.forEach(rule => {
      const groupKey = `${rule.type}s`;
      if (rule.id !== undefined) {
        response[groupKey][rule.id.toString()] = {
          id: rule.id,
          value: rule.value,
          active: rule.active
        };
      }
    });

    return response;
  }

  async updateRulesStatus(ids: NonEmptyList<number>, active: boolean) {
    const existingRules = await this.ruleRepository.findByIds(ids);
    if (existingRules.length !== ids.length) {
      throw new AppError('RULE_NOT_FOUND', 'Specified rule ID not found.', 404);
    }

    const envelope = this.messagePublisher.createEnvelope('UPDATE_RULE', { ids, active });
    await this.messagePublisher.publish(this.queue, envelope);

    const updatedRules = existingRules.map(rule => {
      rule.updateStatus(active);
      return rule.toJSON();
    });

    return {
      updated: updatedRules,
      status: 'success'
    };
  }
}

File: src/application/use-cases/FirewallService.py
import asyncio
from option import Result, Ok, Err
from application_error import ApplicationError
from domain.rule import Rule, RuleType
from application.ports.outbound.rule_repository import RuleRepository
from your_validation_file import ApplicationError


from src.adapters.queue.consumer import AddRuleMessage, DeleteRuleMessage, UpdateStatusMessage


class FirewallService:

    def __init__(self, rule_repository: RuleRepository) -> None:
        self.rule_repository = rule_repository

    async def add_rules(self, command: AddRuleMessage) -> Result[None, ApplicationError]:
        rule_type = command.rule_type
        mode = command.mode

        tasks = []
        for value in command.values:
            # 1. בדיקה מול בסיס הנתונים (כבר היה לנו)
            existing_rule = await self.rule_repository.find_by_type_and_value(rule_type, value)

            if existing_rule:
                if existing_rule.mode != mode:
                    return Err(ApplicationError(
                        code="RULE_CONFLICT",
                        message=f"Cannot add '{value}' to {mode}. It already exists in {existing_rule.mode}."
                    ))
                else:
                    return Err(ApplicationError(
                        code="RULE_ALREADY_EXISTS",
                        message=f"Rule '{value}' already exists in {mode}."
                    ))

            # ==========================================
            # 2. התיקון: שימוש ב-Factory Method של ה-Domain
            # ==========================================
            rule_creation_result = Rule.create(
                rule_type=rule_type,
                mode=mode,
                value=value,
                active=True
            )

            # אם הישות סירבה להיווצר בגלל חוקים עסקיים של הדומיין (כמו IP פגום)
            if rule_creation_result.is_err:
                domain_err = rule_creation_result.unwrap_err()
                # ממירים את שגיאת הדומיין לשגיאת אפליקציה מובנית
                return Err(ApplicationError(
                    code=domain_err.code,
                    message=domain_err.message
                ))

            # שולפים את הישות מתוך ה-Ok
            new_rule = rule_creation_result.unwrap()

            tasks.append(self.rule_repository.save(new_rule))

        await asyncio.gather(*tasks)
        return Ok(None)


    async def delete_rules(self, command: DeleteRuleMessage) -> Result[None, ApplicationError]:
        ids = command.ids

        existing = await self.rule_repository.find_by_ids(ids)
        if len(existing) != len(ids):
            return Err(ApplicationError(
                code="RULE_NOT_FOUND",
                message="Cannot delete. Some rules were not found in the database."
            ))

        await self.rule_repository.delete(ids)
        return Ok(None)


    async def update_rules_status(self, command: UpdateStatusMessage) -> Result[None, ApplicationError]:
        ids = command.ids
        active = command.active

        existing_rules = await self.rule_repository.find_by_ids(ids)
        if len(existing_rules) != len(ids):
            return Err(ApplicationError(
                code="RULE_NOT_FOUND",
                message="Cannot update status. Some rules were not found."
            ))

        for rule in existing_rules:
            rule.update_status(active)

        await self.rule_repository.save_all(existing_rules)
        return Ok(None)